lib/src/metta/runner/builtin_mods/random.rs

```diff
@@ -0,0 +1,292 @@
+    }
+
+    fn reset(&self) {
+        *self.0.borrow_mut() = StdRng::from_os_rng();
+    }
+
+    fn random_range<T, R>(&self, range: R) -> T 
+        where
+            T: rand::distr::uniform::SampleUniform,
+            R: rand::distr::uniform::SampleRange<T>,
+    {
+        self.0.borrow_mut().random_range(range)
+    }
+}
+
+impl Grounded for RandomGenerator {
+    fn type_(&self) -> Atom {
+        ATOM_TYPE_RANDOM_GENERATOR 
+    }
+}
+
+impl Display for RandomGenerator {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        write!(f, "RandomGenerator-{:?}", self.0.as_ptr())
+    }
+}
+
+impl PartialEq for RandomGenerator {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.as_ptr() == other.0.as_ptr()
+    }
+}
+
+
+#[derive(Debug)]
+pub(crate) struct RandomModLoader;
+
+impl ModuleLoader for RandomModLoader {
+    fn load(&self, context: &mut RunContext) -> Result<(), String> {
+        // Initialize module's space
+        let space = GroundingSpace::new();
+        context.init_self_module(space.into(), None);
+
+        // Load module's tokens
+        let _ = self.load_tokens(context.module(), context.metta.clone())?;
+
+        // Parse MeTTa code of the module
+        let parser = SExprParser::new(RANDOM_METTA);
+        context.push_parser(Box::new(parser));
+
+        Ok(())
+    }
+
+    fn load_tokens(&self, target: &MettaMod, _metta: Metta) -> Result<(), String> {
+        let mut tref = target.tokenizer().borrow_mut();
+
+        tref.register_function(GroundedFunctionAtom::new(
+            r"random-int".into(),
+            Atom::expr([ARROW_SYMBOL, ATOM_TYPE_RANDOM_GENERATOR, ATOM_TYPE_NUMBER, ATOM_TYPE_NUMBER, ATOM_TYPE_NUMBER]),
+            random_int));
+
+        tref.register_function(GroundedFunctionAtom::new(
+            r"random-float".into(),
+            Atom::expr([ARROW_SYMBOL, ATOM_TYPE_RANDOM_GENERATOR, ATOM_TYPE_NUMBER, ATOM_TYPE_NUMBER, ATOM_TYPE_NUMBER]),
+            random_float));
+
+        tref.register_function(GroundedFunctionAtom::new(
+            r"set-random-seed".into(),
+            Atom::expr([ARROW_SYMBOL, ATOM_TYPE_RANDOM_GENERATOR, ATOM_TYPE_NUMBER, UNIT_TYPE]),
+            set_random_seed));
+
+        tref.register_function(GroundedFunctionAtom::new(
+            r"new-random-generator".into(),
+            Atom::expr([ARROW_SYMBOL, ATOM_TYPE_NUMBER, ATOM_TYPE_RANDOM_GENERATOR]),
+            new_random_generator));
+
+        tref.register_function(GroundedFunctionAtom::new(
+            r"reset-random-generator".into(),
+            Atom::expr([ARROW_SYMBOL, ATOM_TYPE_RANDOM_GENERATOR, UNIT_ATOM]),
+            reset_random_generator));
+
+        tref.register_function(GroundedFunctionAtom::new(
+            r"flip".into(),
+            Atom::expr([ARROW_SYMBOL, ATOM_TYPE_BOOL]),
+            flip));
+
+        tref.register_token(Regex::new(r"&rng").unwrap(),move |_| { Atom::gnd(RandomGenerator::from_os_rng()) });
```

Comment:
First reason why test fails is that you changed the way `&rng` is instantiated. Before it was a shared link to the same instance:
```
    let generator = RandomGenerator::from_os_rng();
    tref.register_token(regex(r"&rng"), move |_| { Atom::gnd(generator.clone()) });
```
Now it is new instance each time.
```suggestion
    let generator = RandomGenerator::from_os_rng();
    tref.register_token(Regex::new(r"&rng").unwrap(), move |_| { Atom::gnd(generator.clone()) });
```

Reply:
Added this and common code fix in https://github.com/DaddyWesker/hyperon-experimental/pull/20